discord/iterators.py

```diff
@@ -22,20 +22,39 @@
 DEALINGS IN THE SOFTWARE.
 """
 
+from __future__ import annotations
+
 import asyncio
 import datetime
+import sys
+from typing import TYPE_CHECKING, TypeVar, Optional, Any, Callable, Union
 
 from .errors import NoMoreItems
 from .utils import time_snowflake, maybe_coroutine
 from .object import Object
 from .audit_logs import AuditLogEntry
 
+if TYPE_CHECKING:
+    from .member import Member
+    from .user import User
+    from .message import Message
+    from .audit_logs import AuditLogEntry
+    from .guild import Guild
+
+if sys.version_info >= (3, 9):
+    from collections.abc import AsyncIterator, Coroutine
+else:
+    from typing import AsyncIterator, Coroutine
+
+T = TypeVar("T")
+_Predicate = Callable[[T], Union[T, Coroutine[Any, Any, T]]]
```

Comment:
This seems wrong to me.

A predicate is `(T) -> bool` not `(T) -> T`. Granted, it should probably be "convertible to bool" not really strictly `bool`.